akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -442,34 +483,52 @@ private[stream] final case class LimitWeighted[T](n: Long, costFn: T ⇒ Long) e
 /**
  * INTERNAL API
  */
-private[akka] final case class Sliding[T](n: Int, step: Int) extends PushPullStage[T, immutable.Seq[T]] {
-  private var buf = Vector.empty[T]
-
-  override def onPush(elem: T, ctx: Context[immutable.Seq[T]]): SyncDirective = {
-    buf :+= elem
-    if (buf.size < n) {
-      ctx.pull()
-    } else if (buf.size == n) {
-      ctx.push(buf)
-    } else if (step > n) {
-      if (buf.size == step)
-        buf = Vector.empty
-      ctx.pull()
-    } else {
-      buf = buf.drop(step)
-      if (buf.size == n) ctx.push(buf)
-      else ctx.pull()
+private[akka] final case class Sliding[T](n: Int, step: Int) extends GraphStage[FlowShape[T, immutable.Seq[T]]] {
+  require(n > 0, "n must be greater than 0")
+  require(step > 0, "step must be greater than 0")
+
+  val in = Inlet[T]("Sliding.in")
+  val out = Outlet[immutable.Seq[T]]("Sliding.out")
+  override val shape: FlowShape[T, immutable.Seq[T]] = FlowShape(in, out)
+
+  override protected val initialAttributes: Attributes = DefaultAttributes.sliding
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
+    private var buf = Vector.empty[T]
+
+    override def onPush(): Unit = {
+      buf :+= grab(in)
+      if (buf.size < n) {
+        pull(in)
+      } else if (buf.size == n) {
+        push(out, buf)
+      } else if (step <= n) {
+        buf = buf.drop(step)
+        if (buf.size == n) {
+          push(out, buf)
+        } else pull(in)
+      } else if (step > n) {
+        if (buf.size == step) {
```

Comment:
This implementation shares a drawback with the actual one: `.sliding(5, 1000000)` to take 5 samples every million elements will accumulate one million elements in the internal buffer without ever emitting them since they are meant to be skipped.

I played a little to see how it could be done in https://github.com/samueltardieu/akka/blob/graphstage-sliding/akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala#L446-L485.